### ISSUE TITLE
Enable `smart_terminal` functionality outside of the Windows console

### ISSUE THREAD

**Issue by oscarfv** [ninja-build/ninja#2171]
Currently `line_printer.cc` contains this code:

```
#ifndef _WIN32
  smart_terminal_ = isatty(1) && term && string(term) != "dumb";
#else
  if (term && string(term) == "dumb") {
    smart_terminal_ = false;
  } else {
    console_ = GetStdHandle(STD_OUTPUT_HANDLE);
    CONSOLE_SCREEN_BUFFER_INFO csbi;
    smart_terminal_ = GetConsoleScreenBufferInfo(console_, &csbi);
  }
#endif
```

which means in practice that on Windows `smart_terminal` is only set to true when outputting to a console. However, some host applications (like Emacs) are perfectly capable of handling Ninja's one-line progress status and color codes as well.

The problem is how to tell Ninja to behave as if in a smart terminal no matter the absence of a console. I propose to set `smart_terminal` to true when `term` is non-empty and different from "dumb". Ditto for `supports_color`. Is this acceptable? Other suggestions?

**Comment by oscarfv** [comment]
Deciding on the value of `$TERM` alone is not enough, since `ninja` could be executed from an intermediate application that needs to parse its output (CMake is an example) and `$TERM` would be seen by `ninja` as it was on the original environment.

Is it ok to implement a command-line switch or another environment variable  (let's call it `$NINJA_TERM`) to force fancy output?

**Comment by jhasse** [comment]
We need the terminal width for our fancy output. That's why we need `GetConsoleScreenBufferInfo` to work.

**Comment by oscarfv** [comment]
> We need the terminal width for our fancy output. That's why we need `GetConsoleScreenBufferInfo` to work.

What I propose is unrelated to `GetConsoleScreenBufferInfo`. Right now that call is used for obtaining info about the console and, implicitly, to detect that we have a console. I suggest using the value of `$TERM` as an overriding parameter that sets `smart_terminal` to true. On the terminal emulators I tried (and on Emacs) it works fine for enabling the fancy one-line progress meter.

**Comment by jhasse** [comment]
Yes, but later - because `smart_terminal_ == true` -  it is assumed that `GetConsoleScreenBufferInfo` will always return true and we don't check its return value again:
https://github.com/ninja-build/ninja/blob/864f7b5a2f49c764bf3b79bbb3ddd4b989403610/src/line_printer.cc#L82

When forcing `smart_termial_` to true, you'll probably rely on uninitialized memory and `ElideMiddle` won't work.

**Comment by oscarfv** [comment]
That's simple to deal with.

I'll submit a PR as soon as I can.

**Comment by jhasse** [comment]
Is it? How do you plan to get the terminal width?

**Comment by oscarfv** [comment]
Either default to some sensible number (80?) or some very large one. In my experience the terminal emulator does the right thing with the later, albeit it will use the physical next line it the text does not fit in the first one.

**Comment by jhasse** [comment]
Can't you achieve the sams thing by settig NINJA_STATUS to \r?

**Comment by oscarfv** [comment]
Thank you for mentioning `NINJA_STAUTS`, I didn't know that variable and its interesting.

But no, setting it to `\r` does not help because in

https://github.com/ninja-build/ninja/blob/864f7b5a2f49c764bf3b79bbb3ddd4b989403610/src/line_printer.cc#L79

`smart_terminal_` is `false`, so the `else` executes and it is

https://github.com/ninja-build/ninja/blob/864f7b5a2f49c764bf3b79bbb3ddd4b989403610/src/line_printer.cc#L120

which adds a newline.

**Comment by Code7R** [comment]
Would this also fix the unreliable behavior in minTTY (Git-Bash terminal)?

As of now, the behaviour we see on our Windows systems is kinda random. Sometimes the lines are redrawn in place, sometimes not.

Also, sometimes the output is buffered, sometimes not, sometimes buffered in crazy ways (not flushed when needed so you are blind for many seconds, then again flushed randomly, showing a half-written line).
